tests/TestPass.hs

```diff
@@ -0,0 +1,68 @@
+module TestPass (typeCheckPass, fullPassInterp, fullPassJit,
+                 runTestPass, Evaluator, loadTypeEnv) where
+
+import Data.Void
+import Control.Monad.State.Strict
+import qualified Data.Map.Strict as M
+import Unsafe.Coerce
+
+import Pass
+import DeShadow
+import Inference
+import Imp
+import Syntax
+import Type
+import JIT
+import Flops
+import Normalize
+import Interpreter
+import Parser
+import Env
+
+typeCheckPass :: TopPass SourceBlock TopDecl
+typeCheckPass = sourcePass >+> deShadowPass >+> typePass >+> checkTyped
+
+fullPassInterp :: TopPass SourceBlock Void
+fullPassInterp = typeCheckPass >+> interpPass
+
+fullPassJit :: TopPass SourceBlock Void
+fullPassJit = typeCheckPass >+> normalizePass >+> checkNExpr
+                >+> simpPass      >+> checkNExpr
+                >+> impPass       >+> checkImp
+                >+> flopsPass
+                >+> jitPass
+
+
+type TestFullPass env b =  SourceBlock -> TopPassM env b
+
+evalDecl :: Monoid env => TestFullPass env b -> SourceBlock -> StateT env IO ()
+evalDecl pass block = do
+  env <- get
+  (_, env') <- liftIO (runTopPassM env (pass block))
+  modify (<> env')
+
+loadFile :: (Monoid env) => String -> TestFullPass env b -> IO env
+loadFile fname pass = do
+    source <- readFile fname
+    let sourceBlocks = parseProg source
+    execStateT (mapM (evalDecl pass) sourceBlocks) mempty
+
+type Evaluator = SourceBlock -> IO Result'
+
+runTestPass :: String -> TopPass SourceBlock Void -> IO Evaluator
+runTestPass fname (TopPass pass) = do
+    env <- loadFile fname pass
+    let eval source = do
+            ~(Left res, _) <- runTopPassM env (pass source)
+            return res
+    return eval
+
+
+loadTypeEnv :: String -> IO [(SigmaType, Name)]
+loadTypeEnv fname =
+    case sourcePass >+> deShadowPass >+> typePass of
+        TopPass pass -> do
+            envs <- loadFile fname pass
+            let env = (snd (unsafeCoerce envs)) :: TypeEnv
```

Comment:
I am trying to pre-load the definitions from Prelude to the generator environment so it can make use of them in random generation, thus obviating the need to write generators for primitives. The current approach is to evaluate the prelude up to the `typePass` and then extract the [type environment](https://github.com/google-research/dex-lang/blob/674c4df6b212b83b74fdc34885c5b4f8b8d7720b/src/lib/Type.hs#L28). However it is tricky to do the extraction (perhaps need dependent type?). I resorted to the `unsafeCoerce` which is obviously not very satisfying. Is there a better way?

Reply:
We could make versions of the passes that expose their environments rather than hide them. Or we could extract types directly from the parsed program, since all of the names in the prelude have explicit type annotations. But we can do one of those things later. For now, I'm happy with `unsafeCoerce` used here for generating test examples.